File: packages/Thoth.Json.Auto/Prelude.fs
namespace Thoth.Json.Auto

open System
open System.Collections.Generic
open FSharp.Reflection

[<AutoOpen>]
module internal Prelude =

    type FSharpValue with

        static member MakeList(elementType: Type, elements: seq<obj>) : obj =
            let objListType = typedefof<obj list>
            let listType = objListType.MakeGenericType(elementType)
            let ucis = FSharpType.GetUnionCases(listType)
            let emptyUci = ucis |> Seq.find (fun uci -> uci.Name = "Empty")
            let consUci = ucis |> Seq.find (fun uci -> uci.Name = "Cons")
            let empty = FSharpValue.MakeUnion(emptyUci, [||])

            elements
            |> Seq.rev
            |> Seq.fold
                (fun acc x ->
                    FSharpValue.MakeUnion(
                        consUci,
                        [|
                            x
                            acc
                        |]
                    )
                )
                empty

    [<RequireQualifiedAccess>]
    module Map =

        let merge (a: Map<'k, 'v>) (b: Map<'k, 'v>) =
            if a = b then
                a
            else
                seq {
                    for KeyValue(k, v) in a do
                        yield k, v

                    for KeyValue(k, v) in b do
                        yield k, v
                }
                |> Map.ofSeq

    [<RequireQualifiedAccess>]
    module Lazy =

#if FABLE_COMPILER
        let makeGeneric (ty: Type) (func: obj) : obj =
            let factory: unit -> obj = unbox func
            Lazy<obj>(factory) |> box
#else
        open System.Reflection

        type private LazyHelper =
            static member Create<'t>(factory: unit -> 't) : Lazy<'t> =
                Lazy<'t>(factory, true)

        let lazyCreateMethodDefinition =
            typeof<LazyHelper>
                .GetMethods(BindingFlags.Static ||| BindingFlags.NonPublic)
            |> Seq.filter (fun x -> x.Name = "Create")
            |> Seq.exactlyOne
            |> fun mi -> mi.GetGenericMethodDefinition()

        /// Creates a lazy value of the given type.
        /// ty is the type of 't
        /// func is a boxed value of unit -> 't
        let makeGeneric (ty: Type) (func: obj) =
            lazyCreateMethodDefinition
                .MakeGenericMethod([| ty |])
                .Invoke(null, [| func |])
#endif

    [<AutoOpen>]
    module internal ActivePatterns =

        let (|UnitType|_|) (ty: Type) =
            if ty.FullName = typeof<unit>.FullName then
                Some()
            else
                None

        let (|StringType|_|) (ty: Type) =
            if ty.FullName = typeof<string>.FullName then
                Some()
            else
                None

        let (|CharType|_|) (ty: Type) =
            if ty.FullName = typeof<char>.FullName then
                Some()
            else
                None

        let (|IntType|_|) (ty: Type) =
            if ty.FullName = typeof<int>.FullName then
                Some()
            else
                None

        let (|BoolType|_|) (ty: Type) =
            if ty.FullName = typeof<bool>.FullName then
                Some()
            else
                None

        let (|Int64Type|_|) (ty: Type) =
            if ty.FullName = typeof<int64>.FullName then
                Some()
            else
                None

        let (|SingleType|_|) (ty: Type) =
            if ty.FullName = typeof<float32>.FullName then
                Some()
            else
                None

        let (|DoubleType|_|) (ty: Type) =
            if ty.FullName = typeof<float>.FullName then
                Some()
            else
                None

        let (|DecimalType|_|) (ty: Type) =
            if ty.FullName = typeof<decimal>.FullName then
                Some()
            else
                None

        let (|ByteType|_|) (ty: Type) =
            if ty.FullName = typeof<byte>.FullName then
                Some()
            else
                None

        let (|SByteType|_|) (ty: Type) =
            if ty.FullName = typeof<sbyte>.FullName then
                Some()
            else
                None

        let (|UInt16Type|_|) (ty: Type) =
            if ty.FullName = typeof<uint16>.FullName then
                Some()
            else
                None

        let (|Int16Type|_|) (ty: Type) =
            if ty.FullName = typeof<int16>.FullName then
                Some()
            else
                None

        let (|UIntType|_|) (ty: Type) =
            if ty.FullName = typeof<uint>.FullName then
                Some()
            else
                None

        let (|UInt64Type|_|) (ty: Type) =
            if ty.FullName = typeof<uint64>.FullName then
                Some()
            else
                None

        let (|BigIntType|_|) (ty: Type) =
            if ty.FullName = typeof<bigint>.FullName then
                Some()
            else
                None

        let (|GuidType|_|) (ty: Type) =
            if ty.FullName = typeof<Guid>.FullName then
                Some()
            else
                None

        let (|TimeSpanType|_|) (ty: Type) =
            if ty.FullName = typeof<TimeSpan>.FullName then
                Some()
            else
                None

        let (|DateTimeType|_|) (ty: Type) =
            if ty.FullName = typeof<DateTime>.FullName then
                Some()
            else
                None

        let (|DateTimeOffsetType|_|) (ty: Type) =
            if ty.FullName = typeof<DateTimeOffset>.FullName then
                Some()
            else
                None

        let (|OptionType|_|) (ty: Type) =
            if
                ty.IsGenericType
                && ty.GetGenericTypeDefinition() = typedefof<obj option>
            then
                Some(ty.GetGenericArguments()[0])
            else
                None

        let (|ListType|_|) (ty: Type) =
            if
                ty.IsGenericType
                && ty.GetGenericTypeDefinition() = typedefof<obj list>
            then
                Some(ty.GetGenericArguments()[0])
            else
                None

        let (|MapType|_|) (ty: Type) =
            if
                ty.IsGenericType
                && ty.GetGenericTypeDefinition() = typedefof<Map<int, obj>>
            then
                Some(ty.GetGenericArguments()[0], ty.GetGenericArguments()[1])
            else
                None

        let (|ArrayType|_|) (ty: Type) =
#if FABLE_COMPILER
            if ty.IsArray then
                Some(ty.GetElementType())
            else
                None
#else
            if ty.IsArray && ty.GetArrayRank() = 1 then
                Some(ty.GetElementType())
            else
                None
#endif

        let (|SeqType|_|) (ty: Type) =
            if
                ty.IsGenericType
                && ty.GetGenericTypeDefinition() = typedefof<obj seq>
            then
                Some(ty.GetGenericArguments()[0])
            else
                None

        let (|SetType|_|) (ty: Type) =
            if
                ty.IsGenericType
                && ty.GetGenericTypeDefinition() = typedefof<Set<int>>
            then
                Some(ty.GetGenericArguments()[0])
            else
                None

        let (|FSharpRecordType|_|) (ty: Type) =
#if FABLE_COMPILER_PYTHON
            if FSharpType.IsRecord(ty) then
                Some(FSharpType.GetRecordFields(ty))
            else
                None
#else
            if FSharpType.IsRecord(ty, true) then
                Some(
                    FSharpType.GetRecordFields(
                        ty,
                        allowAccessToPrivateRepresentation = true
                    )
                )
            else
                None
#endif

        let (|FSharpUnionType|_|) (ty: Type) =
#if FABLE_COMPILER_PYTHON
            if FSharpType.IsUnion(ty) then
                Some(FSharpType.GetUnionCases(ty))
            else
                None
#else
            if FSharpType.IsUnion(ty, true) then
                Some(
                    FSharpType.GetUnionCases(
                        ty,
                        allowAccessToPrivateRepresentation = true
                    )
                )
            else
                None
#endif

        let (|FSharpTupleType|_|) (ty: Type) =
            if FSharpType.IsTuple(ty) then
                Some(FSharpType.GetTupleElements(ty))
            else
                None

        let (|EnumType|_|) (ty: Type) =
            if ty.IsEnum then
                Some(ty.GetEnumUnderlyingType())
            else
                None

#if !FABLE_COMPILER
        let (|StringEnum|_|) (ty: Type) =
            ty.CustomAttributes
            |> Seq.tryPick (
                function
                | attr when
                    attr.AttributeType.FullName = "Fable.Core.StringEnumAttribute"
                    ->
                    Some attr
                | _ -> None
            )
#endif

#if !FABLE_COMPILER
        let (|CompiledName|_|) (caseInfo: UnionCaseInfo) =
            caseInfo.GetCustomAttributes()
            |> Seq.tryPick (
                function
                | :? CompiledNameAttribute as att -> Some att.CompiledName
                | _ -> None
            )
#endif

#if !FABLE_COMPILER
        let (|LowerFirst|Forward|)
            (args: IList<Reflection.CustomAttributeTypedArgument>)
            =
            args
            |> Seq.tryPick (
                function
                | rule when
                    rule.ArgumentType.FullName = typeof<Fable.Core.CaseRules>
                        .FullName
                    ->
                    Some rule
                | _ -> None
            )
            |> function
                | Some rule ->
                    match rule.Value with
                    | :? int as value ->
                        match value with
                        | 0 -> Forward
                        | 1 -> LowerFirst
                        | _ -> LowerFirst // should not happen
                    | _ -> LowerFirst // should not happen
                | None -> LowerFirst
#endif

    [<RequireQualifiedAccess>]
    module Type =

        let isRecursive (ty: Type) =
            let rec loop (seen: Set<string>) (current: Type) =
                match current with
                | FSharpTupleType elementTypes ->
                    if Array.contains ty elementTypes then
                        true
                    else
                        let seenNext =
                            Set.union
                                seen
                                (elementTypes
                                 |> Array.map (fun ty -> ty.FullName)
                                 |> Set.ofArray)

                        elementTypes
                        |> Seq.filter (fun ty ->
                            not (Set.contains ty.FullName seen)
                        )
                        |> Seq.exists (fun ty -> loop seenNext ty)
                | current when
                    current.IsGenericType
                    && current.GetGenericTypeDefinition() = typedefof<obj list>
                    ->
                    let elementType =
                        current.GetGenericArguments() |> Array.head

                    if elementType = ty then
                        true
                    else
                        let seenNext = Set.add elementType.FullName seen

                        loop seenNext elementType
                | FSharpRecordType fields ->
                    let fieldTypes =
                        fields |> Array.map (fun pi -> pi.PropertyType)

                    if Array.contains ty fieldTypes then
                        true
                    else
                        let seenNext =
                            Set.union
                                seen
                                (fieldTypes
                                 |> Array.map (fun ty -> ty.FullName)
                                 |> Set.ofArray)

                        fieldTypes
                        |> Seq.filter (fun ty ->
                            not (Set.contains ty.FullName seen)
                        )
                        |> Seq.exists (fun ty -> loop seenNext ty)
                | FSharpUnionType fields ->
                    let fieldTypes =
                        fields
                        |> Array.collect (fun uci -> uci.GetFields())
                        |> Array.map (fun pi -> pi.PropertyType)

                    if Array.contains ty fieldTypes then
                        true
                    else
                        let seenNext =
                            Set.union
                                seen
                                (fieldTypes
                                 |> Array.map (fun ty -> ty.FullName)
                                 |> Set.ofArray)

                        fieldTypes
                        |> Seq.filter (fun ty ->
                            not (Set.contains ty.FullName seen)
                        )
                        |> Seq.exists (fun ty -> loop seenNext ty)
                | _ -> false

            loop Set.empty ty

File: assets/live-example/precompiled/86dc26764e27629d/packages/Thoth.Json.Core.Auto/Prelude.js

import { getUnionCaseFields, getEnumUnderlyingType, isEnum, getTupleElements, isTuple, isUnion, getRecordElements, isRecord, getElementType, isArray, class_type, getGenerics, option_type, getGenericTypeDefinition, equals, isGenericType, fullName, makeUnion, name, list_type, obj_type, makeGenericType, getUnionCases } from "fable-library-js/Reflection.js";
import { filter, exists, singleton, collect, append, delay, reverse, fold, find } from "fable-library-js/Seq.js";
import { ofSeq } from "fable-library-js/Map.js";
import { comparePrimitives, structuralHash, Lazy, compare } from "fable-library-js/Util.js";
import { some } from "fable-library-js/Option.js";
import { collect as collect_1, head, map, contains, item } from "fable-library-js/Array.js";
import { empty as empty_1, add, contains as contains_1, ofArray, union } from "fable-library-js/Set.js";

export function Microsoft_FSharp_Reflection_FSharpValue__FSharpValue_MakeList_Static_Z38FAE0DE(elementType, elements) {
    const ucis = getUnionCases(makeGenericType(list_type(obj_type), [elementType]));
    const emptyUci = find((uci) => (name(uci) === "Empty"), ucis);
    const consUci = find((uci_1) => (name(uci_1) === "Cons"), ucis);
    return fold((acc, x) => makeUnion(consUci, [x, acc]), makeUnion(emptyUci, []), reverse(elements));
}

export function Map_merge(a, b) {
    if (a.Equals(b)) {
        return a;
    }
    else {
        return ofSeq(delay(() => append(collect((matchValue) => {
            const activePatternResult = matchValue;
            return singleton([activePatternResult[0], activePatternResult[1]]);
        }, a), delay(() => collect((matchValue_1) => {
            const activePatternResult_1 = matchValue_1;
            return singleton([activePatternResult_1[0], activePatternResult_1[1]]);
        }, b)))), {
            Compare: (x, y) => (compare(x, y) | 0),
        });
    }
}

export function Lazy_makeGeneric(ty, func) {
    return new Lazy(func);
}

export function ActivePatterns_$007CUnitType$007C_$007C(ty) {
    if (fullName(ty) === "Microsoft.FSharp.Core.Unit") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CStringType$007C_$007C(ty) {
    if (fullName(ty) === "System.String") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CCharType$007C_$007C(ty) {
    if (fullName(ty) === "System.Char") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CIntType$007C_$007C(ty) {
    if (fullName(ty) === "System.Int32") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CBoolType$007C_$007C(ty) {
    if (fullName(ty) === "System.Boolean") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CInt64Type$007C_$007C(ty) {
    if (fullName(ty) === "System.Int64") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CSingleType$007C_$007C(ty) {
    if (fullName(ty) === "System.Single") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CDoubleType$007C_$007C(ty) {
    if (fullName(ty) === "System.Double") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CDecimalType$007C_$007C(ty) {
    if (fullName(ty) === "System.Decimal") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CByteType$007C_$007C(ty) {
    if (fullName(ty) === "System.Byte") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CSByteType$007C_$007C(ty) {
    if (fullName(ty) === "System.SByte") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CUInt16Type$007C_$007C(ty) {
    if (fullName(ty) === "System.UInt16") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CInt16Type$007C_$007C(ty) {
    if (fullName(ty) === "System.Int16") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CUIntType$007C_$007C(ty) {
    if (fullName(ty) === "System.UInt32") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CUInt64Type$007C_$007C(ty) {
    if (fullName(ty) === "System.UInt64") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CBigIntType$007C_$007C(ty) {
    if (fullName(ty) === "System.Numerics.BigInteger") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CGuidType$007C_$007C(ty) {
    if (fullName(ty) === "System.Guid") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CTimeSpanType$007C_$007C(ty) {
    if (fullName(ty) === "System.TimeSpan") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CUriType$007C_$007C(ty) {
    if (fullName(ty) === "System.Uri") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CDateTimeType$007C_$007C(ty) {
    if (fullName(ty) === "System.DateTime") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CDateTimeOffsetType$007C_$007C(ty) {
    if (fullName(ty) === "System.DateTimeOffset") {
        return some(undefined);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007COptionType$007C_$007C(ty) {
    if (isGenericType(ty) && equals(getGenericTypeDefinition(ty), option_type(obj_type))) {
        return item(0, getGenerics(ty));
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CListType$007C_$007C(ty) {
    if (isGenericType(ty) && equals(getGenericTypeDefinition(ty), list_type(obj_type))) {
        return item(0, getGenerics(ty));
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CMapType$007C_$007C(ty) {
    if (isGenericType(ty) && equals(getGenericTypeDefinition(ty), class_type("Microsoft.FSharp.Collections.FSharpMap`2", [obj_type, obj_type]))) {
        return [item(0, getGenerics(ty)), item(1, getGenerics(ty))];
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CArrayType$007C_$007C(ty) {
    if (isArray(ty)) {
        return getElementType(ty);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CSeqType$007C_$007C(ty) {
    if (isGenericType(ty) && equals(getGenericTypeDefinition(ty), class_type("System.Collections.Generic.IEnumerable`1", [obj_type]))) {
        return item(0, getGenerics(ty));
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CSetType$007C_$007C(ty) {
    if (isGenericType(ty) && equals(getGenericTypeDefinition(ty), class_type("Microsoft.FSharp.Collections.FSharpSet`1", [obj_type]))) {
        return item(0, getGenerics(ty));
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CFSharpRecordType$007C_$007C(ty) {
    if (isRecord(ty)) {
        return getRecordElements(ty);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CFSharpUnionType$007C_$007C(ty) {
    if (isUnion(ty)) {
        return getUnionCases(ty);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CFSharpTupleType$007C_$007C(ty) {
    if (isTuple(ty)) {
        return getTupleElements(ty);
    }
    else {
        return undefined;
    }
}

export function ActivePatterns_$007CEnumType$007C_$007C(ty) {
    if (isEnum(ty)) {
        return getEnumUnderlyingType(ty);
    }
    else {
        return undefined;
    }
}

export function Type_isRecursive(ty) {
    const loop = (seen_mut, current_mut) => {
        loop:
        while (true) {
            const seen = seen_mut, current = current_mut;
            let current_1;
            const activePatternResult = ActivePatterns_$007CFSharpTupleType$007C_$007C(current);
            if (activePatternResult != null) {
                const elementTypes = activePatternResult;
                if (contains(ty, elementTypes, {
                    Equals: equals,
                    GetHashCode: (x) => (structuralHash(x) | 0),
                })) {
                    return true;
                }
                else {
                    const seenNext = union(seen, ofArray(map(fullName, elementTypes), {
                        Compare: (x_1, y_1) => (comparePrimitives(x_1, y_1) | 0),
                    }));
                    return exists((ty_3) => loop(seenNext, ty_3), filter((ty_2) => !contains_1(fullName(ty_2), seen), elementTypes));
                }
            }
            else if ((current_1 = current, isGenericType(current_1) && equals(getGenericTypeDefinition(current_1), list_type(obj_type)))) {
                const elementType = head(getGenerics(current));
                if (equals(elementType, ty)) {
                    return true;
                }
                else {
                    seen_mut = add(fullName(elementType), seen);
                    current_mut = elementType;
                    continue loop;
                }
            }
            else {
                const activePatternResult_1 = ActivePatterns_$007CFSharpRecordType$007C_$007C(current);
                if (activePatternResult_1 != null) {
                    const fields = activePatternResult_1;
                    const fieldTypes = map((pi) => pi[1], fields);
                    if (contains(ty, fieldTypes, {
                        Equals: equals,
                        GetHashCode: (x_2) => (structuralHash(x_2) | 0),
                    })) {
                        return true;
                    }
                    else {
                        const seenNext_2 = union(seen, ofArray(map(fullName, fieldTypes), {
                            Compare: (x_3, y_3) => (comparePrimitives(x_3, y_3) | 0),
                        }));
                        return exists((ty_6) => loop(seenNext_2, ty_6), filter((ty_5) => !contains_1(fullName(ty_5), seen), fieldTypes));
                    }
                }
                else {
                    const activePatternResult_2 = ActivePatterns_$007CFSharpUnionType$007C_$007C(current);
                    if (activePatternResult_2 != null) {
                        const fields_1 = activePatternResult_2;
                        const fieldTypes_1 = map((pi_1) => pi_1[1], collect_1(getUnionCaseFields, fields_1));
                        if (contains(ty, fieldTypes_1, {
                            Equals: equals,
                            GetHashCode: (x_4) => (structuralHash(x_4) | 0),
                        })) {
                            return true;
                        }
                        else {
                            const seenNext_3 = union(seen, ofArray(map(fullName, fieldTypes_1), {
                                Compare: (x_5, y_5) => (comparePrimitives(x_5, y_5) | 0),
                            }));
                            return exists((ty_9) => loop(seenNext_3, ty_9), filter((ty_8) => !contains_1(fullName(ty_8), seen), fieldTypes_1));
                        }
                    }
                    else {
                        return false;
                    }
                }
            }
            break;
        }
    };
    return loop(empty_1({
        Compare: (x_6, y_6) => (comparePrimitives(x_6, y_6) | 0),
    }), ty);
}

